compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+    files: Array[JFile],
+    flags: Array[String],
+    outDir: JFile
+  ) extends Target {
+    override def toString() = outDir.toString
+  }
+
+  /** A target whose files will be compiled separately according to their
+   *  suffix `_X`
+   */
+  private final case class SeparateCompilationTarget(
+    dir: JFile,
+    flags: Array[String],
+    outDir: JFile
+  ) extends Target {
+
+    /** Get the files grouped by `_X` as a list of groups, files missing this
+     *  suffix will be put into the same group
+     */
+    def compilationUnits: List[Array[JFile]] =
+      dir
+      .listFiles
+      .groupBy { file =>
+        val name = file.getName
+        Try {
+          val potentialNumber = name
+            .substring(0, name.lastIndexOf('.'))
+            .reverse.takeWhile(_ != '_').reverse
+
+          potentialNumber.toInt.toString
+        }
+        .toOption
+        .getOrElse("")
+      }
+      .toList.sortBy(_._1).map(_._2.filter(isCompilable))
+  }
+
+  /** Each `Test` takes the `targets` and performs the compilation and assertions
+   *  according to the implementing class "neg", "run" or "pos".
+   */
+  private abstract class Test(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean) {
+
+    /** Actual compilation run logic, the test behaviour is defined here */
+    protected def compilationRunnable(target: Target): Runnable
+
+    /** All targets left after filtering out */
+    private val allTargets =
+      if (!filter.isDefined) targets
+      else targets.filter {
+        case ConcurrentCompilationTarget(files, _, _) =>
+          files.exists(file => filter.get.findFirstIn(file.getAbsolutePath).isDefined)
+        case SeparateCompilationTarget(dir, _, _) =>
+          filter.get.findFirstIn(dir.getAbsolutePath).isDefined
+      }
+
+    /** Total amount of targets being compiled by this test */
+    val totalTargets = allTargets.length
+
+    private[this] var _errors =  0
+    def errors: Int = synchronized { _errors }
+
+    private[this] var _targetsCompiled = 0
+    private def targetsCompiled: Int = synchronized { _targetsCompiled }
+
+    /** Complete the current compilation with the amount of errors encountered */
+    protected final def completeCompilation(errors: Int) = synchronized {
+      _targetsCompiled += 1
+      _errors += errors
+    }
+
+    private[this] var _failed = false
+    /** Fail the current test */
+    protected[this] final def fail(): Unit = synchronized { _failed = true }
+    def didFail: Boolean = _failed
+
+    /** Instructions on how to reproduce failed target compilations */
+    private[this] val failureInstructions = mutable.ArrayBuffer.empty[String]
+    protected final def addFailureInstruction(ins: String): Unit =
+      synchronized { failureInstructions.append(ins) }
+
+    /** The targets that failed according to the implementing subclass */
+    private[this] val failedCompilationTargets = mutable.ArrayBuffer.empty[String]
```

Comment:
mis-usage of the term "target" makes it really hard to understand what do those methods do